api/src/main/java/org/infinispan/api/async/AsyncStreamingCache.java

```diff
@@ -13,30 +15,67 @@
  */
 public interface AsyncStreamingCache<K> {
    /**
-    * Retrieves the value of the specified key as an {@link InputStream}. It is up to the application to ensure
-    * that the stream is consumed and closed. The marshaller is ignored, i.e. all data will be read in its
-    * raw binary form. The returned input stream implements the {@link CacheEntryMetadata} interface.
-    * The returned input stream is not thread-safe.
+    * Retrieves the value of the specified key as an {@link InputStream}. It is up to the application to ensure that the
+    * stream is consumed and closed. The marshaller is ignored, i.e. all data will be read in its raw binary form. The
+    * returned input stream implements the {@link CacheEntryMetadata} interface. The returned input stream is not
+    * thread-safe.
+    *
+    * @param key key to use
+    */
+   default <T extends InputStream & CacheEntryMetadata> CompletionStage<T> get(K key) {
```

Comment:
> Do we want to add a new interface to encapsulate these types together? Otherwise I fear this is going to be unwieldly to use if the user wants the metadata (requiring casts to get it to work)

I did that because `InputStream` is an abstract class.
 
> Also are we requiring the users to close the stream even if they don't consume it?

Yes, since `InputStream` is `AutoCloseable` the recommended approach is to use try-with-resources